pgjdbc/src/main/java/org/postgresql/util/internal/IntHashSet.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2020, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.util.internal;
+
+import java.util.Arrays;
+import java.util.function.IntConsumer;
+
+/**
+ * A set for any (except -1) int values.
+ * This is an internal class, and it is not meant to be used as a public API.
+ */
+public class IntHashSet {
+  private static final int EMPTY_VALUE = -1;
+
+  private int[] values;
+  private int size;
+
+  public IntHashSet(final int size) {
+    this.values = new int[calculateCapacity(Math.max(4, size << 1))];
+    clear();
+  }
+
+  public int getSize() {
+    return this.size;
+  }
+
+  public void clear() {
+    this.size = 0;
+    Arrays.fill(values, EMPTY_VALUE);
+  }
+
+  public boolean contains(int value) {
+    if (value == EMPTY_VALUE) {
+      illegalValue();
+    }
+    final int[] values = this.values;
+    final int mask = values.length - 1;
+    int index = calculateIndex(value, mask);
+    int v = values[index];
+    while (v != EMPTY_VALUE) {
+      if (v == value) {
+        return true;
+      }
+      index = (index + 1) & mask;
+      v = values[index];
+    }
+    return false;
+  }
+
+  public boolean add(int value) {
+    if (value == EMPTY_VALUE) {
+      illegalValue();
+    }
+    final int[] values = this.values;
+    final int len = values.length;
+    final int mask = len - 1;
+    int index = calculateIndex(value, mask);
+    int v = values[index];
+    while (v != EMPTY_VALUE) {
+      if (v == value) {
+        return false;
+      }
+      index = (index + 1) & mask;
+      v = values[index];
+    }
+    values[index] = value;
+    if (++this.size >= len >> 1) {
+      increaseCapacity();
+    }
+    return true;
+  }
+
+  public boolean remove(int value) {
+    if (value == EMPTY_VALUE) {
+      illegalValue();
+    }
+    final int[] values = this.values;
+    final int mask = values.length - 1;
+    int deleteIndex = calculateIndex(value, mask);
+    int v = values[deleteIndex];
+    while (v != EMPTY_VALUE) {
+      if (v == value) {
+        values[deleteIndex] = EMPTY_VALUE;
+        int index = (deleteIndex + 1) & mask;
+        int vr = values[index];
+        while (vr != EMPTY_VALUE) {
+          int newIndex = calculateIndex(vr, mask);
+          if ((index < newIndex && (newIndex <= deleteIndex || deleteIndex <= index))
+              || (newIndex <= deleteIndex && deleteIndex <= index)) {
+            values[deleteIndex] = vr;
+            values[index] = EMPTY_VALUE;
+            deleteIndex = index;
+          }
+          index = (index + 1) & mask;
+          vr = values[index];
+        }
+        this.size--;
+        return true;
+      }
+      deleteIndex = (deleteIndex + 1) & mask;
+      v = values[deleteIndex];
```

Comment:
This is tricky. Could you document it?
Do you think this could be simplified at a cost of reduced performance?

I guess the performance of `.remove(..)` is not that important for pgjdbc, so it would probably be better to value code readability.